localstack/services/s3/v3/provider.py

```diff
@@ -0,0 +1,1231 @@
+    ) -> GetObjectOutput:
+        # TODO: missing handling parameters:
+        #  if_match: IfMatch = None,
+        #  if_modified_since: IfModifiedSince = None,
+        #  if_none_match: IfNoneMatch = None,
+        #  if_unmodified_since: IfUnmodifiedSince = None,
+        #  response_cache_control: ResponseCacheControl = None,
+        #  response_content_disposition: ResponseContentDisposition = None,
+        #  response_content_encoding: ResponseContentEncoding = None,
+        #  response_content_language: ResponseContentLanguage = None,
+        #  response_content_type: ResponseContentType = None,
+        #  response_expires: ResponseExpires = None,
+        #  request_payer: RequestPayer = None,
+        #  part_number: PartNumber = None,
+        #  expected_bucket_owner: AccountId = None,
+
+        store = self.get_store(context.account_id, context.region)
+        bucket_name = request["Bucket"]
+        object_key = request["Key"]
+        if not (s3_bucket := store.buckets.get(bucket_name)):
+            raise NoSuchBucket("The specified bucket does not exist", BucketName=bucket_name)
+
+        # TODO implement PartNumber once multipart is done (being able to select only a Part)
+
+        s3_object = s3_bucket.get_object(
+            key=object_key,
+            version_id=request.get("VersionId"),
+            http_method="GET",
+        )
+
+        response = GetObjectOutput(
+            AcceptRanges="bytes",
+            **s3_object.get_system_metadata_fields(),
+        )
+        if s3_object.user_metadata:
+            response["Metadata"] = s3_object.user_metadata
+
+        if s3_object.parts and request.get("PartNumber"):
+            response["PartsCount"] = len(s3_object.parts)
+
+        if s3_object.version_id:
+            response["VersionId"] = s3_object.version_id
+
+        if s3_object.website_redirect_location:
+            response["WebsiteRedirectLocation"] = s3_object.website_redirect_location
+
+        if checksum_algorithm := s3_object.checksum_algorithm:
+            # this is a bug in AWS: it sets the content encoding header to an empty string (parity tested)
+            response["ContentEncoding"] = ""
+            if request.get("ChecksumMode", "").upper() == "ENABLED":
+                response[f"Checksum{checksum_algorithm.upper()}"] = s3_object.checksum_value
+
+        s3_stored_object = self._storage_backend.open(bucket_name, s3_object)
+
+        if range_header := request.get("Range"):
+            range_data = parse_range_header(range_header, s3_object.size)
+            s3_stored_object.seek(range_data.begin)
+            response["Body"] = LimitedIterableStream(
+                s3_stored_object, max_length=range_data.content_length
+            )
+            response["ContentRange"] = range_data.content_range
+            response[
+                "ContentLength"
+            ] = range_data.content_length  # TODO: should we set it for chunked encoding?
```

Comment:
```suggestion
            # TODO: should we set it for chunked encoding?
            response["ContentLength"] = range_data.content_length  
```

in werkzeug, the `Transfer-Encoding: chunked` header will be set automatically if the content-length is not set. if the content length is set, and you want it to be chunked, you need to also set the header explicitly.

i was told that #8841 is relevant here ;-)

Reply:
Yes, now I know we always need to return it 😄